File: backend/app/domain/models/api_key.py
from typing import Optional, List
from datetime import datetime, UTC
from pydantic import BaseModel
from enum import Enum


class APIKeyScope(str, Enum):
    FULL = "full"


class APIKeyStatus(str, Enum):
    ACTIVE = "active"
    REVOKED = "revoked"


class APIKey(BaseModel):
    id: str
    user_id: str
    name: str
    key_prefix: str
    key_hash: str
    scopes: List[APIKeyScope] = [APIKeyScope.FULL]
    status: APIKeyStatus = APIKeyStatus.ACTIVE
    expires_at: Optional[datetime] = None
    last_used_at: Optional[datetime] = None
    created_at: datetime = datetime.now(UTC)
    updated_at: datetime = datetime.now(UTC)

    def is_valid(self) -> bool:
        if self.status != APIKeyStatus.ACTIVE:
            return False
        if self.expires_at and datetime.now(UTC) > self.expires_at:
            return False
        return True

    def revoke(self):
        self.status = APIKeyStatus.REVOKED
        self.updated_at = datetime.now(UTC)

    def update_last_used(self):
        self.last_used_at = datetime.now(UTC)
        self.updated_at = datetime.now(UTC)

File: frontend/src/api/apiKey.ts
import { apiClient, ApiResponse } from './client';

export type APIKeyScope = 'full';
export type APIKeyStatus = 'active' | 'revoked';

export interface APIKey {
  id: string;
  name: string;
  key_prefix: string;
  scopes: APIKeyScope[];
  status: APIKeyStatus;
  expires_at: string | null;
  last_used_at: string | null;
  created_at: string;
}

export interface CreateAPIKeyResponse extends APIKey {
  key: string;
}

export interface CreateAPIKeyRequest {
  name: string;
  scopes: APIKeyScope[];
  expires_in_days: number | null;
}

export async function listAPIKeys(): Promise<APIKey[]> {
  const response = await apiClient.get<ApiResponse<APIKey[]>>('/api-keys');
  return response.data.data;
}

export async function createAPIKey(request: CreateAPIKeyRequest): Promise<CreateAPIKeyResponse> {
  const response = await apiClient.post<ApiResponse<CreateAPIKeyResponse>>('/api-keys', request);
  return response.data.data;
}

export async function revokeAPIKey(keyId: string): Promise<void> {
  await apiClient.delete<ApiResponse<{}>>(`/api-keys/${keyId}`);
}
